plugins/catalog-backend-module-msgraph/src/microsoftGraph/read.ts

```diff
@@ -574,25 +412,104 @@ export async function readMicrosoftGraphOrg(
       logger: options.logger,
     });
-    users.push(...usersWithFilter);
+    users = usersWithFilter;
+  } else {
+    options.logger?.info(
+      'Not loading users as neither user or userGroupMember configuration has been provided',
+    );
+    users = [];
   }
-  const { groups, rootGroup, groupMember, groupMemberOf } =
-    await readMicrosoftGraphGroups(client, tenantId, {
+
+  let groups: GroupEntity[];
+  if (options.groupSearch || options.groupFilter) {
+    const {
+      groups: groups2,
+      rootGroup,
+      groupMember,
+      groupMemberOf,
+    } = await readMicrosoftGraphGroups(client, tenantId, {
       queryMode: options.queryMode,
       groupSearch: options.groupSearch,
       groupFilter: options.groupFilter,
       groupSelect: options.groupSelect,
       groupTransformer: options.groupTransformer,
       organizationTransformer: options.organizationTransformer,
+      logger: options.logger,
     });
 
-  resolveRelations(rootGroup, groups, users, groupMember, groupMemberOf);
+    groups = groups2;
+
+    options.logger?.info('Resolving relations');
+    resolveRelations(rootGroup, groups, users, groupMember, groupMemberOf);
+  } else {
+    options.logger?.info(
+      'Not loading groups as no group configuration has been provided',
+    );
+    groups = [];
+  }
+
   users.sort((a, b) => a.metadata.name.localeCompare(b.metadata.name));
   groups.sort((a, b) => a.metadata.name.localeCompare(b.metadata.name));
 
+  if (users.length === 0 && groups.length === 0) {
+    options.logger?.warn(
+      'Microsoft Graph Entity Provider loaded no users or groups. Check your configuration.',
+    );
+  }
+
   return { users, groups };
 }
 
+async function transformUsers(
+  client: MicrosoftGraphClient,
+  users: Iterable<User> | AsyncIterable<User>,
+  logger?: Logger,
+  transformer?: UserTransformer,
+) {
+  const limiter = limiterFactory(10);
+
+  const resolvedTransformer = transformer ?? defaultUserTransformer;
+  const promises: Promise<void>[] = [];
+  const entities: UserEntity[] = [];
+
+  // Process all users in parallel, otherwise it can take quite some time
+  for await (const user of users) {
+    promises.push(
+      limiter(async () => {
+        let userPhoto;
+        try {
+          userPhoto = await client.getUserPhotoWithSizeLimit(
+            user.id!,
+            // We are limiting the photo size, as users with full resolution photos
+            // can make the Backstage API slow
+            120,
+          );
+        } catch (e) {
+          logger?.warn(`Unable to load user photo for`, {
+            user: user.id,
+            error: e,
+          });
+        }
+
+        const entity = await resolvedTransformer(user, userPhoto);
+
+        if (entity) {
+          entities.push(entity);
+        }
+      }),
+    );
+  }
+
+  // Wait for all users and photos to be downloaded
+  await Promise.all(promises);
+
+  logger?.info('Finished transforming users', {
```

Comment:
this feels more like a debug level thing

Reply:
Fixed in #15891